### DISCUSSION TITLE
Comparison to raw ProseMirror, esp. regarding collaboration.

### DISCUSSION THREAD

**Discussion by tslocke** [ueberdosis/tiptap#1507] [General Questions]
Hi,

I'm in the process of choosing between tiptap, and using prosemirror directly.

Has anything being written up on the pros and cons? I get the impression that the main goal of tiptap is a better developer experience. Is that right? Does it come at a cost, or is all the underlying power of ProseMirror still available?

I'm particularly interested in collaboration. ProseMirror already has very solid support for collaboration, but it seems that tiptap does things a bit differently using Y.js. What problem does this solve? It seems like a lot of work which makes me think there must be some major shortcomings to ProseMirror collaboration.

My app will likely need to use the ProseMirror transforms/transactions API directly, e.g. users can interactively revert/apply edits from other users. I'm concerned the Y.js collaboration model makes that stuff more implicit or "magic". In my case I don't want collaborative editing to "just work", I need access to the mechanics.

This comment in the docs is of interest:

> The underyling schema tiptap uses is an excellent foundation to sync documents

My understanding is that ProseMirror allows collaborative editing for /any/ schema. Are there some gotchas?

Any advice much appreciated.

Tom

**Comment by philippkuehn** [comment]
Hey, using the y.js collab extension is absolutely optional. Of course you can write your own tiptap extension to use `prosemirror-collab` ([there was one in v1](https://github.com/ueberdosis/tiptap/blob/v1/packages/tiptap-extensions/src/extensions/Collaboration.js) which got replaced by the y.js extension in v2). Basically you only have to wrap the ProseMirror plugin in a tiptap extension. But the biggest part of using `prosemirror-collab` is the [server integration](https://github.com/ProseMirror/website/tree/master/src/collab/server).

Everything you can do with ProseMirror, you can also do with tiptap. The goal of tiptap is to bundle some related logic into extensions (e.g. schema, commands, input rules), and to make some complex APIs more accessible (e.g. chained commands). We also provide a documentation with many examples and extensions so it’s way easier to start than with plain ProseMirror.

**Reply by philippkuehn** [reply]
y.js builds on top of another technology (CRDT instead of OT). there are some interesting reads why this solution is that great :)
https://josephg.com/blog/crdts-are-the-future/

**Reply by tslocke** [reply]
Yeah I know that's the key difference. What I wanted to clarify was if CRDTs allow you to work with change-sets as first class values, e.g. to apply/revert them in different orders.

Was just reading up on automerge and it looks like this is indeed possible:

> Automerge keeps track of the changes you make to the state, so that you can view old versions, compare versions, create branches, and choose when to merge them.
(Similar to git, which allows diffing, branching, merging, and pull request workflows.)

Not sure if Y.js has such features. Thanks for the advice.

**Reply by philippkuehn** [reply]
There is an example for creating versions. Maybe that helps?

https://docs.yjs.dev/ecosystem/editor-bindings/prosemirror#versions-and-showing-the-differences

**Reply by tslocke** [reply]
Awesome - from the README in the ProseMirror + Versions Demo:

> The snapshot API is very versatile and allows you to render the differences between any two snapshots. In this examples, we simply created a linear version history. But using the same API you can model branching & merging as well.

https://github.com/yjs/yjs-demos/tree/main/prosemirror-versions

Thanks!

**Reply by hanspagel** [reply]
I have to chime in here. From what I know Kevin doesn’t think of this as production-ready. I have no idea what the short comings are though, I didn’t even looked at the code yet. 

That said, there are multiple teams working on that already, so I’m pretty positive there will be a community effort *or* a way to collect enough funding so we (Kevin + us) can take that to a production-grade level at some point.

Anyway, we’re doing a big bet on Y.js. CRDT are the future, and Y.js is the most performant implementation. It has a huge ecosystem, and new community projects pop up every week. If you want to go the OT-way, you can do that with tiptap as you would do it with ProseMirror. All parts and APIs of ProseMirror are accessible through tiptap.

Regarding all your other questions: ProseMirror is fine, we love it. We won’t add a “con” list to our docs. This is something someone wrote me yesterday, maybe it helps:

> We're trying to build documents based on a fairly complex schema with custom components and we've hit the wall several times using other editors (Quill / Slate). It was fairly straightforward to transition from Slate to Tiptap based on the documentation you have already. I actually prefer your decision to prioritize documentation via examples over detailed API documentation. ProseMirror has detailed documentation, but is quite imposing to someone new. I think one of the best aspects of Tiptap is that it makes ProseMirror approachable.

**Reply by tslocke** [reply]
@hanspagel "doesn't think of this as production-ready" - which piece of the puzzle specifically did you mean? prosemirror-versions in yjs?

Thanks for the feedback

**Reply by hanspagel** [reply]
Yes! https://demos.yjs.dev/prosemirror-versions/prosemirror-versions.html

Not very qualified opinion, just wanted to say as a far as I know it’s not ready. If that’s important to you, slide into my emails humans@tiptap.dev, I’ll take this to my next call with Kevin and let you know then.
